### ISSUE TITLE
asyncssh Library Logs Using Root Logger When Installed From Python Egg

### ISSUE THREAD

**Issue by jannero** [ronf/asyncssh#156]
Meaning that you can not set silence asyncssh's logging

The following code demonstrates the issue:

```
import logging
import asyncssh.logging

root_logger = logging.getLogger()

def print_root_logger_level():
    print(logging.getLevelName(root_logger.level))

print_root_logger_level()

asyncssh.logging.set_log_level(logging.DEBUG)

print_root_logger_level()

print(asyncssh.logger)
```

The output is
```
WARNING
DEBUG
<_SSHLogger root (DEBUG)>
```

So the default level of root logger is WARNING and after setting log level using asyncssh.logging.set_log_level to DEBUG also the root logger level is changed.

Also all the messages logged by asyncssh seems to be logged through the root logger.

Relevan version and environment:
asyncssh 1.13.2
Python 3.7
Archlinux

I wasn't able to reproduce this issue when running with asyncssh from Git, i.e. with
```
python -m venv venv
venv/bin/python setup.py develop
```

I then packaged asyncssh into wheel myself with `python setup.py bdist_wheel`. When installed from the wheel the above test does what it is supposed to do so the fix would be just to distribute asyncssh as wheel instead of egg.

The `asyncssh.logging` uses `__package__` to define the used logger and I'm guessing that when the code is run from Python egg the `__package__` is an empty string. I did not find confirmation for this with a quick search though

**Comment by ronf** [comment]
Thanks for the report!

So far, I haven't been able to reproduce this here on either MacOS or Linux, even when installing asyncssh via pip. Were there any special options you gave when installing it?

Also, if you change __package__ to __name__ in logging.py, do you still see the problem? I'd like to try and find something that could work with both egg and wheel packaging, as well as when run in unpacked form from an sdist or from git.

**Comment by jannero** [comment]
Ok, it seems to work if you just run `pip install asyncssh`.

Actually in my case I had asyncssh installed because it is a dependency of another project.

The difference is that if you install asyncssh with `pip install asyncssh` it will extract the files, i.e. in the site-packages directory there is `asyncss` directory.

If you install asyncssh as a dependency, the site-packages directory contains the egg file and then you can see the incorrect behaviour.

So with the following `setup.py`:

```python
import setuptools

setuptools.setup(
    name='asyncssh-logging-test',
    py_modules='asyncssh_logging_example',
    install_requires=['asyncssh'],
    setup_requires=['setuptools_autover'],
)
```

and `asyncssh_logging_example.py`:
```python
import logging
import asyncssh.logging

root_logger = logging.getLogger()

def print_root_logger_level():
    print(logging.getLevelName(root_logger.level))

print_root_logger_level()

asyncssh.logging.set_log_level(logging.DEBUG)

print_root_logger_level()

print(asyncssh.logger)
```

So here is the test (run in a directory where the above two python files are):

```sh
# Note that I used my system's python this time (now 3.6 instead of 3.7 I first used); the behaviour is the same as with Python 3.7
$ python --version
Python 3.6.6

$ python -m venv venv
$ venv/bin/pip install --quiet asyncssh
$ venv/bin/pip list
Package      Version
------------ -------
asn1crypto   0.24.0
asyncssh     1.13.2
cffi         1.11.5
cryptography 2.3
idna         2.7
pip          10.0.1
pycparser    2.18
setuptools   39.0.1
six          1.11.0

# The example works as expected
$ venv/bin/python asyncssh_logging_example.py
WARNING
WARNING
<_SSHLogger asyncssh (DEBUG)>

# Here you can see that if you install asyncssh directly the egg is extracted
$ ls -l venv/lib/python3.6/site-packages/ | grep asyncssh
drwxr-x--- 4 5 asyncssh
drwxr-x--- 2 5 asyncssh-1.13.2-py3.6.egg-info

$ rm -rf venv/

$ python -m venv venv
$ venv/bin/python setup.py install >/dev/null 2>/dev/null
$ venv/bin/pip list
Package               Version
--------------------- -------
asn1crypto            0.24.0
asyncssh              1.13.2
asyncssh-logging-test 0.0.0
cffi                  1.11.5
cryptography          2.3
idna                  2.7
pip                   10.0.1
pycparser             2.18
setuptools            39.0.1
six                   1.11.0

# Now the issue is visible
$ venv/bin/python asyncssh_logging_example.py
WARNING
DEBUG
<_SSHLogger root (DEBUG)>

# And now the egg is not extracted
$ ls -l venv/lib/python3.6/site-packages/ | grep asyncssh
-rw-r----- 1 janne janne 552819 Jul 21 07:57 asyncssh-1.13.2-py3.6.egg
-rw-r----- 1 janne janne   1007 Jul 21 07:57 asyncssh_logging_test-0.0.0-py3.6.egg
```

I did the following change:
```diff
diff --git a/asyncssh/logging.py b/asyncssh/logging.py
index 1d0e74c..128d08c 100644
--- a/asyncssh/logging.py
+++ b/asyncssh/logging.py
@@ -20,7 +20,7 @@ class _SSHLogger(logging.LoggerAdapter):
     """Adapter to add context to AsyncSSH log messages"""
 
     _debug_level = 1
-    _pkg_logger = logging.getLogger(__package__)
+    _pkg_logger = logging.getLogger("asyncssh")
 
     def __init__(self, parent=_pkg_logger, child=None, context=''):
         self._context = context
```

And created egg with `python setup.py bdist_egg` and copied the egg from *dist* directory into the virtualenv's site-packages that contained the egg installed by `python setup.py install` in the above example, i.e. I overwrote the installed egg with the one I just created myself and the issue is gone (as expected).

**Comment by ronf** [comment]
Thanks very much for the detailed reproduction instructions and the suggested fix! I've been able to reproduce the problem here. While it's not ideal to hard-code "asyncssh" instead of using "__package__" to figure that out, it does seem like the simplest fix given the bug of "__package__" not working right with eggs.

I've gone ahead and checked this change into the develop branch, and will include it in the next release.

**Comment by ronf** [comment]
I don't think wheels were around when I started the project, or if they were they were brand new and the popular distribution format at the time was still eggs.

Since then, I've just never really spent the time to look into switching. Is it well supported these days on all platforms (including Windows)?

**Comment by jannero** [comment]
Wheels have been around since 2012: https://packaging.python.org/discussions/wheel-vs-egg/

Also you could check this https://pythonwheels.com

**Comment by jannero** [comment]
.. And I wasn't very clear in my first post that you could fix the issue just by moving to wheels from egg.

**Comment by ronf** [comment]
Reading up on wheels at the links above as well as an article at http://lucumr.pocoo.org/2014/1/27/python-on-wheels/ I think you're right  - it seems like I should try distributing a wheel to PyPI in place of the egg I'm building today. In addition to fixing the issue you ran into here, it appears to provide some other nice benefits, and it will still support a "universal" (pure Python) wheel, which was one of my concerns. In other places I've seen wheels used, there was usually binary content involved.

If I'm able to get this to work, I'll probably back out the other change and go back to using __package__ before I do the next release.

**Comment by ronf** [comment]
Ok - converting to wheels was straightforward, so I've gone ahead and released AsyncSSH 1.13.3 with the previous change reverted as a wheel instead of an egg. Thanks again for helping to identify this issue and for your suggestions about how to resolve it!

If you run into any further problems, please let me know.
